### PR TITLE
[CAS-1185] Refac - Retry connection

### PR DESCRIPTION
**Pull request by leandroBorgesFerreira**
[CAS-1185](https://stream-io.atlassian.net/browse/CAS-1185)

### 🎯 Goal

Create a retry logic when the connection fails in a nonpermanent way.

### 🛠 Implementation details

When there's no internet and the socket executes an error handling, it retries 3 times to reconnect with exponential backoff. 

### 🧪 Testing

Enable and disable internet while using the app

### ☑️ Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] PR targets the `develop` branch
- ~[ ] Changelog is updated with client-facing changes~
- [x] New code is covered by unit tests
- ~[ ] Comparison screenshots added for visual changes~
- ~[ ] Affected documentation updated (docusaurus, tutorial, CMS (task created))~
- [x] Reviewers added

### 🎉 GIF

giphy

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt [hunk 1]
```diff
@@ -122,7 +127,13 @@ internal class ChatSocketServiceImpl constructor(
             ChatErrorCode.UNABLE_TO_PARSE_SOCKET_EVENT.code,
             ChatErrorCode.NO_ERROR_BODY.code,
             -> {
-                // Nothing to do on that case
+                if (!networkStateProvider.isConnected() && reconnectionAttempts < 3) {
```
Comment (ogkuzmin) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:130]: It's better not to expand the complexity of this class by this logic. We already have `HealthMonitor` with similar logic, we can modify it to cover our case.

Comment (ogkuzmin) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:130]: Move this magic number to constants

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt [hunk 2]
```diff
@@ -72,6 +76,7 @@ internal class ChatSocketServiceImpl constructor(
                     }
                     is State.Connected -> {
                         healthMonitor.start()
+                        reconnectionAttempts = 0
```
[the remark this answers is not in the document]
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:79]: 🤔  Do you mean logout from the user? This prevents retries stop happening if a user:
1 - Retries 3 times.
2 - Successfully connects
3 - Loses connection again. 

If we only reset in a logout, then this case is not covered.

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt (hunk 1 above)
Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:130]: `by this logic` do you mean `!networkStateProvider.isConnected()`?

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (ogkuzmin) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt:81]: We don't need to restore retries count when connected. We need this only for first connection
I agree we can leave the case with logout

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt (hunk 1 above)
Comment (ogkuzmin) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt]: Also retries condition

Comment (leandroBorgesFerreira) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/ChatSocketServiceImpl.kt]: This way `HealthMonitor` will start to be too much specific and handle logic looks outside of its scope.

**Review by ogkuzmin — APPROVED** [review]
LGTM for me now!

**Pull request merged**